core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -114,119 +114,143 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
-                    otherDecimal[otherOffset + 1],
+                    decimal[decimalOffset],
+                    decimal[decimalOffset + 1],
+                    otherDecimal[otherDecimalOffset],
+                    otherDecimal[otherDecimalOffset + 1],
                     decimal,
-                    offset);
-            state.addOverflow(overflow + otherState.getOverflow());
+                    decimalOffset);
+            counterOverflow[counterOverflowOffset] += overflow + otherCounterOverflow[otherCounterOverflowOffset + 1];
         }
         else {
-            state.setNotNull();
-            decimal[offset] = otherDecimal[otherOffset];
-            decimal[offset + 1] = otherDecimal[otherOffset + 1];
-            state.setOverflow(otherState.getOverflow());
+            counterOverflowState.setNotNull();
+            decimalState.setNotNull();
+            decimal[decimalOffset] = otherDecimal[otherDecimalOffset];
+            decimal[decimalOffset + 1] = otherDecimal[otherDecimalOffset + 1];
+            counterOverflow[counterOverflowOffset + 1] = otherCounterOverflow[otherCounterOverflowOffset + 1];
         }
     }
 
-    public static void outputShortDecimal(DecimalType type, LongDecimalWithOverflowAndLongState state, BlockBuilder out)
+    public static void outputShortDecimal(DecimalType type, Int128State decimalState, Int128State counterOverflowState, BlockBuilder out)
     {
-        if (state.getLong() == 0) {
+        long[] counterOverflow = counterOverflowState.getDecimalArray();
+        int counterOverflowOffset = counterOverflowState.getDecimalArrayOffset();
+
+        // No ale jak. A co jezeli nie bylo wierszy, to nie moze byc null
+        if (!counterOverflowState.isNotNull()) {
             out.appendNull();
         }
         else {
-            writeShortDecimal(out, average(state, type).toLongExact());
+            writeShortDecimal(out, average(decimalState, counterOverflowState, type).toLongExact());
         }
     }
 
-    public static void outputLongDecimal(DecimalType type, LongDecimalWithOverflowAndLongState state, BlockBuilder out)
+    public static void outputLongDecimal(DecimalType type, Int128State decimalState, Int128State counterOverflowState, BlockBuilder out)
     {
-        if (state.getLong() == 0) {
+        long[] counterOverflow = counterOverflowState.getDecimalArray();
+        int counterOverflowOffset = counterOverflowState.getDecimalArrayOffset();
+
+        if (counterOverflow[counterOverflowOffset] == 0) {
             out.appendNull();
         }
         else {
-            type.writeObject(out, average(state, type));
+            type.writeObject(out, average(decimalState, counterOverflowState, type));
         }
     }
 
     @VisibleForTesting
-    public static Int128 average(LongDecimalWithOverflowAndLongState state, DecimalType type)
+    public static Int128 average(Int128State decimalState, Int128State counterOverflowState, DecimalType type)
```

Comment:
Are we gonna be able to skip overflow serialization where all values are 0 if we use Int128State for both counter and overflow?
Can we use two separate `LongState`s (it may have perf impact)?